File: dynamic/louti.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution{
public:
    int test(int n){
        vector<vector<int>> dp(n+1, vector<int>(2));
        dp[1][0] = 1;
        dp[2][0] = 1;
        dp[2][1] = 1;
        dp[3][0] = 2;
        dp[3][1] = 2;

        for(int i = 4; i < n + 1; i++){
            dp[i][0] = dp[i - 1][0] + dp[n-1][1];
            if(i > 3){
                dp[i][1] = dp[n - 2][0] + dp[n - 2][0] + dp[n - 3][0]; 
            }
            
        }

        return dp[n][0] + dp[n][1];
    }
};


int main(){
    Solution * solu = new Solution();
    cout << solu->test(4);
    cin.get();
    return 0;
}

File: bishi/7_24/louti.py


from calendar import Calendar


class Solution:
    def test(self, n):

        if(n == 1):
            return 1
        if(n == 2):
            return 2
        if(n == 3):
            return 4

        

        f = [1] * (n + 1)
        f[1] = 1
        f[2] = 1
        f[3] = 3
        f[4] = 6

        for i in range(4, n):
            f[i] = f[i - 1]
            if(i >= 4):
                f[i] += f[i - 4]

        
        print(f[n - 1])
        print(f[n - 2])
        print(f[n - 3])
        print(f[n - 4])

        f[n] = f[n -1] + f[n -2] + f[n -3]

        return f[n]


if __name__ == "__main__":
    s = Solution()
    print(s.test(4))